## Step 1: Import relevant packages

In [1]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neighbors import NearestNeighbors
from sklearn.datasets import load_iris
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import SGDClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import train_test_split

import numpy as np
import matplotlib.pyplot as plt
import cv2
from tensorflow import keras

2022-10-12 03:03:37.351531: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


## Step 2: Load the Fashion-MNIST dataset

In [2]:
(x_train, y_train), (x_test, y_test) = keras.datasets.fashion_mnist.load_data()

## Step 3: Take a subset of the data set

In [3]:
# Initially, take 3,000 samples for training and 1,000 samples for testing.
x_dataset = np.concatenate((x_train, x_test), axis=0)
y_dataset = np.concatenate((y_train, y_test), axis=0)

X_train, X_test, Y_train, Y_test = train_test_split(x_dataset, y_dataset, test_size=0.2, random_state=0)

X_train3000 = X_train[:3000]
Y_train3000 = Y_train[:3000]
X_test1000 = X_test[:1000]
Y_test1000 = Y_test[:1000]
print(X_train3000.shape)
print(X_test1000.shape)
print(Y_train3000.shape)
print(Y_test1000.shape)

(3000, 28, 28)
(1000, 28, 28)
(3000,)
(1000,)


In [4]:
# double the number of training samples from 3,000 to 6,000
X_train6000 = X_train[:6000]
Y_train6000 = Y_train[:6000]
print(X_train6000.shape)
print(Y_train6000.shape)

(6000, 28, 28)
(6000,)


In [5]:
# then from 6,000 to 12,000, 
X_train12000 = X_train[:12000]
Y_train12000 = Y_train[:12000]
print(X_train12000.shape)
print(Y_train12000.shape)

(12000, 28, 28)
(12000,)


## Step 4: Perform necessary reshaping of the data for the classifiers

In [6]:
X_train3000 = X_train3000.reshape(X_train3000.shape[0], -1)
X_test1000 = X_test1000.reshape(X_test1000.shape[0], -1)
print(X_train3000.shape)
print(X_test1000.shape)

(3000, 784)
(1000, 784)


In [7]:
# double the number of training samples from 3,000 to 6,000
X_train6000 = X_train6000.reshape(X_train6000.shape[0], -1)
print(X_train6000.shape)

(6000, 784)


In [8]:
# then from 6,000 to 12,000, 
X_train12000 = X_train12000.reshape(X_train12000.shape[0], -1)
print(X_train12000.shape)

(12000, 784)


## KNN

Initially, take 3,000 samples for training and 1,000 samples for testing.

In [10]:
from sklearn import metrics
# Step 5: Initialise the classifier model
knn_classifier = KNeighborsClassifier(n_neighbors=3)
# Step 6: Fit the model to the training data
knn_classifier.fit(X_train3000, Y_train3000)
# Step 7: Use the trained/fitted model to evaluate the testing data
knn_pred = knn_classifier.predict(X_test1000)
# Step 8: Report the performance of each classifier
accuracy_knn = metrics.accuracy_score(Y_test1000, knn_pred)
precision_knn = metrics.precision_score(Y_test1000, knn_pred, average='macro')
recall_knn = metrics.recall_score(Y_test1000, knn_pred, average = 'macro')
f1_score_knn = metrics.f1_score(Y_test1000, knn_pred, average = 'macro')
confusion_matrix_knn = metrics.confusion_matrix(Y_test1000, knn_pred)

print(f'Accuracy of KNN: {accuracy_knn}')
print(f'Precision of KNN: {precision_knn}')
print(f'Recall of KNN: {recall_knn}')
print(f'F1 score of KNN: {f1_score_knn}')
print(f'Confusion matrix of KNN:\n {confusion_matrix_knn}')

Accuracy of KNN: 0.778
Precision of KNN: 0.7874788219427613
Recall of KNN: 0.7773691052384246
F1 score of KNN: 0.7750910928717593
Confusion matrix of KNN:
 [[ 84   0   1   2   0   0  11   0   1   0]
 [  0 111   1   2   1   0   0   0   0   0]
 [  6   0  76   0  20   1   8   0   0   0]
 [ 13   4   5  64   6   0   1   0   1   0]
 [  3   0  20   1  62   0  15   0   0   0]
 [  2   0   0   0   0  74   0  17   0  12]
 [ 22   0  14   1  10   0  45   0   0   0]
 [  0   0   0   0   0   0   0  74   0   3]
 [  2   0   3   0   0   1   2   2  96   0]
 [  1   0   1   0   0   0   0   6   0  92]]


6,000 samples for training and 1,000 samples for testing.

In [11]:
# Step 6: Fit the model to the training data
knn_classifier.fit(X_train6000, Y_train6000)
# Step 7: Use the trained/fitted model to evaluate the testing data
knn_pred = knn_classifier.predict(X_test1000)
# Step 8: Report the performance of each classifier
accuracy_knn = metrics.accuracy_score(Y_test1000, knn_pred)
precision_knn = metrics.precision_score(Y_test1000, knn_pred, average='macro')
recall_knn = metrics.recall_score(Y_test1000, knn_pred, average = 'macro')
f1_score_knn = metrics.f1_score(Y_test1000, knn_pred, average = 'macro')
confusion_matrix_knn = metrics.confusion_matrix(Y_test1000, knn_pred)

print(f'Accuracy of KNN: {accuracy_knn}')
print(f'Precision of KNN: {precision_knn}')
print(f'Recall of KNN: {recall_knn}')
print(f'F1 score of KNN: {f1_score_knn}')
print(f'Confusion matrix of KNN:\n {confusion_matrix_knn}')

Accuracy of KNN: 0.81
Precision of KNN: 0.8163858384302187
Recall of KNN: 0.8094840713615868
F1 score of KNN: 0.8080998363570335
Confusion matrix of KNN:
 [[ 86   0   1   2   0   0   9   0   1   0]
 [  2 110   1   1   0   0   1   0   0   0]
 [  6   0  81   0  16   1   7   0   0   0]
 [  9   2   5  74   1   0   2   0   1   0]
 [  1   0  18   0  67   0  15   0   0   0]
 [  1   0   0   0   0  76   0  13   0  15]
 [ 20   0  14   0   8   0  50   0   0   0]
 [  0   0   0   0   0   2   0  73   0   2]
 [  2   0   0   1   1   0   2   1  99   0]
 [  0   0   1   0   0   0   1   4   0  94]]


12,000 samples for training and 1,000 samples for testing.

In [12]:
# Step 6: Fit the model to the training data
knn_classifier.fit(X_train12000, Y_train12000)
# Step 7: Use the trained/fitted model to evaluate the testing data
knn_pred = knn_classifier.predict(X_test1000)
# Step 8: Report the performance of each classifier
accuracy_knn = metrics.accuracy_score(Y_test1000, knn_pred)
precision_knn = metrics.precision_score(Y_test1000, knn_pred, average='macro')
recall_knn = metrics.recall_score(Y_test1000, knn_pred, average = 'macro')
f1_score_knn = metrics.f1_score(Y_test1000, knn_pred, average = 'macro')
confusion_matrix_knn = metrics.confusion_matrix(Y_test1000, knn_pred)

print(f'Accuracy of KNN: {accuracy_knn}')
print(f'Precision of KNN: {precision_knn}')
print(f'Recall of KNN: {recall_knn}')
print(f'F1 score of KNN: {f1_score_knn}')
print(f'Confusion matrix of KNN:\n {confusion_matrix_knn}')

Accuracy of KNN: 0.816
Precision of KNN: 0.8196580092812894
Recall of KNN: 0.8147083841235665
F1 score of KNN: 0.8129048889643513
Confusion matrix of KNN:
 [[ 81   1   1   2   0   0  13   0   1   0]
 [  0 111   1   1   1   0   1   0   0   0]
 [  4   0  87   0   9   0  10   1   0   0]
 [  9   2   3  74   3   0   2   0   1   0]
 [  1   0  16   2  70   0  12   0   0   0]
 [  0   0   0   0   0  78   0  14   0  13]
 [ 19   0  13   0  12   0  48   0   0   0]
 [  0   0   0   0   0   0   0  74   0   3]
 [  1   1   0   0   1   0   2   1  99   1]
 [  0   0   1   1   0   0   0   4   0  94]]


### Our result is less accurate than that on the paper. The knn of the paper is about 0.85 and 0.95, but I only get about 0.78, 0.81, 0.82. This may be because we only set three neighbors here. The number is too small for this dataset, and maybe n = 5 or 9 is better. Also, I find that we do more training sample, get better results.

## SGD

Initially, take 3,000 samples for training and 1,000 samples for testing.

In [13]:
# Step 5: Initialise the classifier model
sgd_classifier = SGDClassifier(max_iter=250)
# Step 6: Fit the model to the training data
sgd_classifier.fit(X_train3000, Y_train3000)
# Step 7: Use the trained/fitted model to evaluate the testing data
sgd_pred = sgd_classifier.predict(X_test1000)
# Step 8: Report the performance of each classifier
accuracy_sgd = metrics.accuracy_score(Y_test1000, sgd_pred)
precision_sgd = metrics.precision_score(Y_test1000, sgd_pred, average='macro')
recall_sgd = metrics.recall_score(Y_test1000, sgd_pred, average = 'macro')
f1_score_sgd = metrics.f1_score(Y_test1000, sgd_pred, average = 'macro')
confusion_matrix_sgd = metrics.confusion_matrix(Y_test1000, sgd_pred)

print(f'Accuracy of SGD: {accuracy_sgd}')
print(f'Precision of SGD: {precision_sgd}')
print(f'Recall of SGD: {recall_sgd}')
print(f'F1 score of SGD: {f1_score_sgd}')
print(f'Confusion matrix of SGD:\n {confusion_matrix_sgd}')

Accuracy of SGD: 0.781
Precision of SGD: 0.7822483110826501
Recall of SGD: 0.7795498982661585
F1 score of SGD: 0.7742511149236707
Confusion matrix of SGD:
 [[ 80   0   0   5   1   0  10   0   3   0]
 [  1 111   0   2   0   0   1   0   0   0]
 [  6   0  58   2  28   1  13   0   3   0]
 [  7   0   3  77   4   0   2   0   1   0]
 [  2   1   6   3  81   0   8   0   0   0]
 [  0   0   0   0   2  80   0   8   7   8]
 [ 17   0   8   4  26   0  34   0   3   0]
 [  0   0   0   0   0   4   0  68   1   4]
 [  3   0   0   0   1   1   1   1  99   0]
 [  0   0   0   0   0   3   1   1   2  93]]


6,000 samples for training and 1,000 samples for testing.

In [14]:
# Step 6: Fit the model to the training data
sgd_classifier.fit(X_train6000, Y_train6000)
# Step 7: Use the trained/fitted model to evaluate the testing data
sgd_pred = sgd_classifier.predict(X_test1000)
# Step 8: Report the performance of each classifier
accuracy_sgd = metrics.accuracy_score(Y_test1000, sgd_pred)
precision_sgd = metrics.precision_score(Y_test1000, sgd_pred, average='macro')
recall_sgd = metrics.recall_score(Y_test1000, sgd_pred, average = 'macro')
f1_score_sgd = metrics.f1_score(Y_test1000, sgd_pred, average = 'macro')
confusion_matrix_sgd = metrics.confusion_matrix(Y_test1000, sgd_pred)

print(f'Accuracy of SGD: {accuracy_sgd}')
print(f'Precision of SGD: {precision_sgd}')
print(f'Recall of SGD: {recall_sgd}')
print(f'F1 score of SGD: {f1_score_sgd}')
print(f'Confusion matrix of SGD:\n {confusion_matrix_sgd}')

Accuracy of SGD: 0.762
Precision of SGD: 0.7919274985601015
Recall of SGD: 0.7598601731733755
F1 score of SGD: 0.7570723063163154
Confusion matrix of SGD:
 [[ 86   0   0   2   2   0   7   0   2   0]
 [  4 107   1   1   2   0   0   0   0   0]
 [  6   0  50   0  49   1   5   0   0   0]
 [ 10   0   2  67  14   0   1   0   0   0]
 [  2   0   6   0  88   0   5   0   0   0]
 [  1   0   0   0   2  88   0   5   2   7]
 [ 14   1   9   1  46   0  21   0   0   0]
 [  0   0   0   0   0   3   0  67   1   6]
 [  2   0   0   0   6   1   6   0  91   0]
 [  0   0   0   0   0   2   0   1   0  97]]


12,000 samples for training and 1,000 samples for testing.

In [15]:
# Step 6: Fit the model to the training data
sgd_classifier.fit(X_train12000, Y_train12000)
# Step 7: Use the trained/fitted model to evaluate the testing data
sgd_pred = sgd_classifier.predict(X_test1000)
# Step 8: Report the performance of each classifier
accuracy_sgd = metrics.accuracy_score(Y_test1000, sgd_pred)
precision_sgd = metrics.precision_score(Y_test1000, sgd_pred, average='macro')
recall_sgd = metrics.recall_score(Y_test1000, sgd_pred, average = 'macro')
f1_score_sgd = metrics.f1_score(Y_test1000, sgd_pred, average = 'macro')
confusion_matrix_sgd = metrics.confusion_matrix(Y_test1000, sgd_pred)

print(f'Accuracy of SGD: {accuracy_sgd}')
print(f'Precision of SGD: {precision_sgd}')
print(f'Recall of SGD: {recall_sgd}')
print(f'F1 score of SGD: {f1_score_sgd}')
print(f'Confusion matrix of SGD:\n {confusion_matrix_sgd}')

Accuracy of SGD: 0.797
Precision of SGD: 0.7989197540163577
Recall of SGD: 0.7954874834828334
F1 score of SGD: 0.7958031207259427
Confusion matrix of SGD:
 [[ 68   0   2   5   3   0  17   0   4   0]
 [  2 109   0   1   2   0   1   0   0   0]
 [  2   0  74   1  20   1  12   0   1   0]
 [  2   0   2  81   7   0   2   0   0   0]
 [  0   0  15   2  73   1  10   0   0   0]
 [  0   0   0   0   0  86   1   7   7   4]
 [  9   0  14   4  12   0  50   0   3   0]
 [  0   0   0   0   0   3   0  68   1   5]
 [  1   0   1   0   0   1   3   0 100   0]
 [  0   0   0   0   0   5   0   6   1  88]]


### Our results are worse than those on paper. The accuracy of this paper is close to 0.819 in the fashion industry, but I only get about 0.78(3000 training),0.76(6000 training). Maybe because we only use max_Iter is set to 250 to limit the maximum number of algorithm iterations. Also the training times of our model are far less than those of the model in this paper, which may lead to insufficient fitting and reduced accuracy. So when we use 12000 samples training, we got very close to paper's accuracy.

## Decision Tree

Initially, take 3,000 samples for training and 1,000 samples for testing.

In [19]:
# Step 5: Initialise the classifier model
dt_classifier = DecisionTreeClassifier()
# Step 6: Fit the model to the training data
dt_classifier.fit(X_train3000, Y_train3000)
# Step 7: Use the trained/fitted model to evaluate the testing data
dt_pred = dt_classifier.predict(X_test1000)
# Step 8: Report the performance of each classifier
accuracy_dt = metrics.accuracy_score(Y_test1000, dt_pred)
precision_dt = metrics.precision_score(Y_test1000, dt_pred, average='macro')
recall_dt = metrics.recall_score(Y_test1000, dt_pred, average = 'macro')
f1_score_dt = metrics.f1_score(Y_test1000, dt_pred, average = 'macro')
confusion_matrix_dt = metrics.confusion_matrix(Y_test1000, dt_pred)

print(f'Accuracy of SGD: {accuracy_dt}')
print(f'Precision of SGD: {precision_dt}')
print(f'Recall of SGD: {recall_dt}')
print(f'F1 score of SGD: {f1_score_dt}')
print(f'Confusion matrix of SGD:\n {confusion_matrix_dt}')

Accuracy of SGD: 0.723
Precision of SGD: 0.723770031801464
Recall of SGD: 0.7204559909585355
F1 score of SGD: 0.7212921159148163
Confusion matrix of SGD:
 [[ 61   0   8   3   1   0  25   0   1   0]
 [  1 107   0   4   1   0   2   0   0   0]
 [  7   2  63   0  21   2  16   0   0   0]
 [  5   6   2  69   7   0   4   0   1   0]
 [  2   1  16   8  53   1  17   0   2   1]
 [  0   0   0   0   0  86   0  10   4   5]
 [ 14   1  14   4  11   0  46   0   2   0]
 [  0   0   0   0   0   7   0  62   1   7]
 [  1   2   4   1   0   0   2   1  93   2]
 [  0   0   2   0   0   6   0   9   0  83]]


6,000 samples for training and 1,000 samples for testing.

In [20]:
# Step 6: Fit the model to the training data
dt_classifier.fit(X_train6000, Y_train6000)
# Step 7: Use the trained/fitted model to evaluate the testing data
dt_pred = dt_classifier.predict(X_test1000)
# Step 8: Report the performance of each classifier
accuracy_dt = metrics.accuracy_score(Y_test1000, dt_pred)
precision_dt = metrics.precision_score(Y_test1000, dt_pred, average='macro')
recall_dt = metrics.recall_score(Y_test1000, dt_pred, average = 'macro')
f1_score_dt = metrics.f1_score(Y_test1000, dt_pred, average = 'macro')
confusion_matrix_dt = metrics.confusion_matrix(Y_test1000, dt_pred)

print(f'Accuracy of SGD: {accuracy_dt}')
print(f'Precision of SGD: {precision_dt}')
print(f'Recall of SGD: {recall_dt}')
print(f'F1 score of SGD: {f1_score_dt}')
print(f'Confusion matrix of SGD:\n {confusion_matrix_dt}')

Accuracy of SGD: 0.738
Precision of SGD: 0.7378112331168756
Recall of SGD: 0.7371170216109972
F1 score of SGD: 0.736686816025544
Confusion matrix of SGD:
 [[ 70   2   4   7   2   0  10   0   4   0]
 [  1 105   0   3   1   1   4   0   0   0]
 [  2   1  63   2  16   2  21   0   2   2]
 [  7   6   3  69   4   1   4   0   0   0]
 [  2   2  18   8  57   0  12   0   2   0]
 [  0   1   0   0   0  88   0   7   4   5]
 [ 10   0   8   4  16   1  51   0   2   0]
 [  0   0   0   0   0   7   0  65   0   5]
 [  3   0   3   1   1   1   4   1  91   1]
 [  0   0   0   3   1   9   1   7   0  79]]


12,000 samples for training and 1,000 samples for testing.

In [21]:
# Step 6: Fit the model to the training data
dt_classifier.fit(X_train12000, Y_train12000)
# Step 7: Use the trained/fitted model to evaluate the testing data
dt_pred = dt_classifier.predict(X_test1000)
# Step 8: Report the performance of each classifier
accuracy_dt = metrics.accuracy_score(Y_test1000, dt_pred)
precision_dt = metrics.precision_score(Y_test1000, dt_pred, average='macro')
recall_dt = metrics.recall_score(Y_test1000, dt_pred, average = 'macro')
f1_score_dt = metrics.f1_score(Y_test1000, dt_pred, average = 'macro')
confusion_matrix_dt = metrics.confusion_matrix(Y_test1000, dt_pred)

print(f'Accuracy of SGD: {accuracy_dt}')
print(f'Precision of SGD: {precision_dt}')
print(f'Recall of SGD: {recall_dt}')
print(f'F1 score of SGD: {f1_score_dt}')
print(f'Confusion matrix of SGD:\n {confusion_matrix_dt}')

Accuracy of SGD: 0.744
Precision of SGD: 0.7519454637909927
Recall of SGD: 0.7405433119928151
F1 score of SGD: 0.744420032856023
Confusion matrix of SGD:
 [[ 67   0   1   3   4   0  23   0   1   0]
 [  1 107   0   6   1   0   0   0   0   0]
 [  4   2  63   2  19   1  19   0   1   0]
 [  5   4   4  66   9   1   4   0   1   0]
 [  1   1  18   8  59   0  13   0   1   0]
 [  0   0   0   0   0  96   0   2   2   5]
 [ 11   0   8   6  21   1  44   0   1   0]
 [  1   0   0   0   0  11   0  61   0   4]
 [  4   1   1   0   0   1   8   0  89   2]
 [  0   0   0   0   0   3   1   4   0  92]]


### We found that the more sample training data, the better accuracy we got. However, our results are less accuracy than its on paper. The accuracy by DT classifier in paper is between 0.78 and 0.79. But our accuracy is  0.723 (3000 training), 0.738(6000 training), 0.744(12000 training). The reason may be that the subset we get from the whole dataset may not have the same data distribution as the whole dataset used in the paper, which affects the accuracy of the training classifier.